crypto/evp/kdf_lib.c

```diff
@@ -0,0 +1,176 @@
+
+/* This array needs to be in order of NIDs */
+static const EVP_KDF_METHOD *standard_methods[] = {
+    &pbkdf2_kdf_meth,
+#ifndef OPENSSL_NO_SCRYPT
+    &scrypt_kdf_meth,
+#endif
+    &tls1_prf_kdf_meth,
+    &hkdf_kdf_meth,
+};
+
+DECLARE_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                           kmeth);
+
+static int kmeth_cmp(const EVP_KDF_METHOD *const *a,
+                     const EVP_KDF_METHOD *const *b)
+{
+    return ((*a)->nid - (*b)->nid);
+}
+
+IMPLEMENT_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                             kmeth);
+
+static const EVP_KDF_METHOD *kdf_meth_find(int type)
+{
+    EVP_KDF_METHOD tmp;
+    const EVP_KDF_METHOD *t = &tmp, **ret;
+
+    tmp.nid = type;
+    ret = OBJ_bsearch_kmeth(&t, standard_methods,
+                            OSSL_NELEM(standard_methods));
+    if (ret == NULL || *ret == NULL)
+        return NULL;
+    return *ret;
+}
+
+EVP_KDF_CTX *EVP_KDF_CTX_new_id(int id)
+{
+    EVP_KDF_CTX *ret;
+    const EVP_KDF_METHOD *kmeth;
+
+    kmeth = kdf_meth_find(id);
+    if (kmeth == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_NEW_ID, EVP_R_UNSUPPORTED_ALGORITHM);
+        return NULL;
+    }
+
+    ret = OPENSSL_zalloc(sizeof(*ret));
+    if (ret == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_NEW_ID, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    ret->kmeth = kmeth;
+
+    if (kmeth->init && kmeth->init(ret) <= 0) {
+        ret->kmeth = NULL;
+        EVP_KDF_CTX_free(ret);
+        return NULL;
+    }
+
+    return ret;
+}
+
+void EVP_KDF_CTX_free(EVP_KDF_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+    if (ctx->kmeth && ctx->kmeth->cleanup)
+        ctx->kmeth->cleanup(ctx);
+    OPENSSL_free(ctx);
+}
+
+int EVP_KDF_CTX_ctrl(EVP_KDF_CTX *ctx, int cmd, int p1, void *p2)
+{
+    int ret;
+
+    if (ctx == NULL || ctx->kmeth == NULL || ctx->kmeth->ctrl == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);
+        return -2;
+    }
+
+    ret = ctx->kmeth->ctrl(ctx, cmd, p1, p2);
+
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);
+
+    return ret;
+}
+
+int EVP_KDF_CTX_ctrl_uint64(EVP_KDF_CTX *ctx, int cmd, uint64_t value)
+{
+    return EVP_KDF_CTX_ctrl(ctx, cmd, 0, &value);
+}
+
+int EVP_KDF_CTX_ctrl_str(EVP_KDF_CTX *ctx,
+                         const char *name, const char *value)
+{
+    if (ctx == NULL || ctx->kmeth == NULL || ctx->kmeth->ctrl_str == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL_STR, EVP_R_COMMAND_NOT_SUPPORTED);
+        return -2;
+    }
+    return ctx->kmeth->ctrl_str(ctx, name, value);
+}
+
+/* Utility functions to send a string or hex string to a ctrl */
+
+int EVP_KDF_CTX_str2ctrl(EVP_KDF_CTX *ctx, int cmd, const char *str)
+{
+    size_t len;
+
+    len = strlen(str);
+    if (len > INT_MAX)
+        return -1;
+    return ctx->kmeth->ctrl(ctx, cmd, len, (void *)str);
+}
```

Comment:
Yes. I will fix that.

Reply:
I have removed NULL checks of pointers that should never be NULL.
All `EVP_KDF_CTX` objects have a method table, so I don't check if kmeth is NULL.
All KDFs require `ctrl` and `derive` methods, otherwise there is no way to provide any inputs or get output.
The `ctrl_str` and `output_size` methods are optional, so they are NULL-checked.